astropy/utils/state.py

```diff
@@ -7,7 +7,79 @@
 __all__ = ['ScienceState']
 
 
-class ScienceState:
+class StateMixinMeta(type):
+    """Metaclass for state information mixin.
+
+    Creates the ``_state`` variable, if doesn't already exist.
+    The namespace is created by the ``__prepare__`` method,
+    so ``_state`` is defined before even a variable definition.
+
+    items in ``_state`` can be accessed as items of the class.
+
+    Examples
+    --------
+    ::
+
+        class MyClass(metaclass=StateMixinMeta):
+            _state["foo"] = "bar"
+
+        assert MyClass["foo"] == "bar"
+        True
+
+    """
+
+    @classmethod
+    def __prepare__(cls, *args):
+        namespace = dict(_state=dict())
+        return namespace
+
+    def __getitem__(cls, key):
+        """
+        Items in ``_state`` can be accessed as items of the class.
+        """
+        return cls._state[key]
+
+
+class _Context(metaclass=StateMixinMeta):
+    """
+    Science state context. Returned by :func:`~ScienceState.set`.
+
+    ::
+
+        class MyState(ScienceState):
+            _value = None
+
+        old_context = MyState.set("test")
+
+        assert old_context._value is None
+        assert MyState._value == "test"
+
+    """
+
+    def __init__(self, parent, value):
+        self._value = value
+        self._parent = parent
+
+    def __enter__(self):
+        """Enter ``with`` statement."""
+        pass
+
+    def __exit__(self, type, value, tb):
+        """Exit ``with`` statement, resetting value and clearing context."""
+        self._parent.set(self._value)  # reset
+        self.__dict__.clear()  # delete on exit
```

Comment:
Is this to remove any possible references? Though wouldn't generally the context itself go out of scope so be cleaned up?

Reply:
Yes. As it currently stands, the `_Context.__enter__` does not return anything, so it would just go out of scope and be cleaned up. However, if the ``__enter__`` method returns the context for use in `with` statements, this is needed.

```
MyState._value = 1
with MyState.set(value=2) as context:
    pass
```

Although, as noted above `_Context` would need to be slightly reconfigured to redirect `_value` etc to `self._parent` because they store the old state.

Reply:
This will be in the next PR.